pytest_allclose/plugin.py

```diff
@@ -0,0 +1,220 @@
+            # the other array)
+            close[[i - 1, -i]] = True
+
+        result = np.all(close)
+
+        if print_fail > 0 and not result:
+            diffs = []
+            # broadcast a and b to have same shape as close for indexing
+            broadcast_a = a + np.zeros(b.shape, dtype=a.dtype)
+            broadcast_b = b + np.zeros(a.shape, dtype=b.dtype)
+            for k, ind in enumerate(zip(*(~close).nonzero())):
+                if k >= print_fail:
+                    break
+                diffs.append("%s: %s %s" % (
+                    ind, broadcast_a[ind], broadcast_b[ind]))
+
+            print("allclose first %d failures:\n  %s" % (
+                len(diffs), "\n  ".join(diffs)))
+
+        return result
+
+    return _allclose
+
+
+_allclose_arg_types = dict(
+    atol=float,
+    rtol=float,
+    xtol=int,
+    equal_nan=bool,
+    print_fail=int,
+    record_rmse=bool,
+)
+
+
+def _rms(x, axis=None, keepdims=False):
+    return np.sqrt(np.mean(x**2, axis=axis, keepdims=keepdims))
+
+
+def _safe_rms(x):
+    x = np.asarray(x)
+    return _rms(x).item() if x.size > 0 else np.nan
+
+
+def _get_allclose_overrides(request):
+    nodename = request.node.nodeid
+    tol_cfg = request.config.inicfg.get("allclose_test_tolerances", "")
+
+    # multiple overrides will match subsequent ``allclose`` calls
+    overrides = []
+
+    for line in (x for x in tol_cfg.split('\n') if len(x) > 0):
+        # each line contains a pattern and list of kwargs (e.g. `atol=0.1`)
+        split_line = line.split()
+        pattern = "*" + split_line[0]
+
+        # escape special characters in pattern
+        escaped = "[]?"
+        pattern = list(pattern)
+        for i, c in enumerate(pattern):
+            if c in escaped:
+                pattern[i] = "[%s]" % c
+        pattern = "".join(pattern)
+
+        if fnmatch(nodename, pattern):
+            kwargs = {}
+            for entry in split_line[1:]:
+                if entry.startswith("#"):
+                    break
+
+                k, v = entry.split("=")
+                if k not in _allclose_arg_types:
+                    raise ValueError("Unrecognized argument %r" % k)
+
+                kwargs[k] = _allclose_arg_types[k](v)
+
+            overrides.append(kwargs)
+
+    return overrides
+
+
+def report_rmses(terminalreporter, relative=True):
+    """Report RMSEs recorded by the allclose fixture in the Pytest terminal.
+
+    This function helps with reporting recorded root mean squared errors.
+
+    Parameters
+    ----------
+    terminalreporter :
+        The terminal reporter object provided by ``pytest_terminal_summary``.
+    relative : bool (optional)
+        Whether to print relative (default) or absolute RMSEs. Relative RMSEs
+        are normalized by the mean RMS of ``a`` and ``b`` in ``allclose``.
+
+    Examples
+    --------
+
+    In your ``conftest.py`` file, put the following to report
+
+    ..code-block:: python
```

Comment:
In general, I find it tricky, because I call `help` a lot on functions/classes, and I like to have the answers right there in the docstring. But for an example like this, maybe it makes sense just to have it in the docs and link to it.

Reply:
I reduced the duplication in code but still have it in both docstrings now using a simple decorator. I also added a test of the docstring of the function you get, while the docstring for the fixture is in the docs.

The only place that it looks bad is in the output of `pytest --fixtures`:

```
--------------------- fixtures defined from pytest_allclose.plugin ---------------------
allclose
    Returns a function checking if two arrays are close, mimicking `numpy.allclose`.
    
    .. currentmodule:: allclose
    
    .. function:: _allclose(a, b, rtol=1e-5, atol=1e-8, xtol=0, equal_nan=False,                             print_fail=5, record_rmse=True)
       :noindex:
    
    Parameters
    ----------
    a : np.ndarray
        First array to be compared.
    b : np.ndarray
        Second array to be compared.
    rtol : float, optional
        Relative tolerance between a and b (relative to b).
    atol : float, optional
        Absolute tolerance between a and b.
    xtol : int, optional
        Allow signals to be right or left shifted by up to ``xtol``
        indices along the first axis
    equal_nan : bool, optional
        If True, nans will be considered equal to nans.
    print_fail : int, optional
        If > 0, print out the first ``print_fail`` entries failing
        the allclose check along the first axis.
    record_rmse : bool, optional
        Whether to record the RMSE value for this comparison. Defaults to True.
        Set to False whenever ``a`` and ``b`` should be far apart
        (when ensuring two signals are sufficiently different, for example).
    
    Returns
    -------
    bool
        True if the two arrays are considered close according to the tolerances.

```

While the ReST directives help with the output in the docs, they are a bit of an eyesore here. I do expect that 95% of the views will be on the website docs and only 5% on source code or `pytest --fixtures`, but it would still be nice to be able to do this better somehow... I think this is an okay compromise right now though.